daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1129,28 +1131,36 @@ private[lf] object SBuiltin {
                 IE.ContractNotActive(coid, expectedTemplateIdOpt.getOrElse(ifaceId), nid)
               )
             )
-          checkTemplateId(cached.templateId) {
-            machine.returnValue = cached.value
-          }
+          if (
+            machine.compiledPackages
+              .getDefinition(ImplementsDefRef(cached.templateId, ifaceId))
+              .isEmpty
+          )
+            throw SErrorDamlException(
+              IE.ContractDoesNotImplementInterface(
+                interfaceId = ifaceId,
+                coid = coid,
+                templateId = cached.templateId,
+              )
+            )
```

Comment:
I'm not very happy with changing the order of this error relative to `WronglyTypedContract`...

Reply:
I tend to agree with Sofia here. Unfortunately, I think that means doing the type check inside the guard becomes difficult. I think the guard should run after this.  I think this might actually be the reason why we didn’t do it in the guard initiially.